src/backend/cdb/cdbpartition.c

```diff
@@ -3856,20 +3854,15 @@ selectListPartition(PartitionNode *partnode, Datum *values, bool *isnull,
 						 * So lhstypid = type(expr) and rhstypeid = type(rule)
 						 */
 
-						/* The tupdesc tuple descriptor matches the table schema, so it has the rule type */
-						Oid rhstypid = tupdesc->attrs[attno - 1]->atttypid;
-
 						/*
-						 * exprTypeOid is passed to us from our caller which evaluated the expression.
-						 * In some cases (e.g legacy optimizer doing explicit casting), we don't compute
-						 * specify exprTypeOid.
-						 * Assume lhstypid = rhstypid in those cases
+						 * Type of rule is in fact the type of const inside the
+                         * rule object.
+                         * Type of expr is defined by attribute in tuple
+                         * descriptor whose number corresponds to current
+                         * partitioning key in partitioned table
 						 */
-						Oid lhstypid = exprTypeOid;
-						if (!OidIsValid(lhstypid))
-						{
-							lhstypid = rhstypid;
-						}
+						Oid rhstypid = c->consttype;
+						Oid lhstypid = tupdesc->attrs[attno - 1]->atttypid;
```

Comment:
The same question as in https://github.com/arenadata/gpdb/pull/167#discussion_r623690992